### PR TITLE
ddl: fix unique key constraint check for 'create table'

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

In MySQL:

```
mysql> create table part7 (col1 int not null,col2 int not null,col3 int not null unique,unique key(col1, col2)) partition by range (col1 + col2) (partition p1 values less than (11), partition p2 values less than (15));
ERROR 1503 (HY000): A PRIMARY KEY must include all columns in the table's partitioning function
```

In TiDB:
```
mysql> create table part7 (col1 int not null,col2 int not null,col3 int not null unique,unique key(col1, col2)) partition by range (col1 + col2) (partition p1 values less than (11), partition p2 values less than (15));
Query OK, 0 rows affected (0.04 sec)
```

### What is changed and how it works?

Every unique key on the table must use every column in the table's partitioning expression.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @ciscoxll @zimulala @winkyao

File: ddl/partition.go [hunk 1]
```diff
@@ -431,56 +431,53 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
-func extractConstraintsColumnNames(cons []*ast.Constraint) []map[string]struct{} {
-	var constraints []map[string]struct{}
-	for _, v := range cons {
-		if v.Tp == ast.ConstraintUniq || v.Tp == ast.ConstraintPrimaryKey {
+	// Every unique key on the table must use every column in the table's partitioning expression.
+	// See https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html
+	for _, constraint := range constraints {
+		switch constraint.Tp {
+		case ast.ConstraintPrimaryKey, ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex:
 			uniKeys := make(map[string]struct{})
```
Comment (ngaut) [inline ddl/partition.go]: Crate map with capacity

Comment (tiancaiamao) [reply ddl/partition.go]: Done
This map is unnecessary

**Comment by tiancaiamao** [comment]
PTAL @ciscoxll @winkyao @zimulala

**Comment by tiancaiamao** [comment]
PTAL @zimulala  @winkyao @crazycs520

**Comment by tiancaiamao** [comment]
PTAL @zimulala @crazycs520 @lysu

File: ddl/partition.go [hunk 2]
```diff
@@ -435,56 +435,36 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
+	// Every unique key on the table must use every column in the table's partitioning expression.
+	// See https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html
+	for _, constraint := range constraints {
+		switch constraint.Tp {
+		case ast.ConstraintPrimaryKey, ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex:
+			if !checkUniqueKeyIncludePartKey(partCols, constraint.Keys) {
+				if constraint.Tp == ast.ConstraintUniq {
+					return ErrUniqueKeyNeedAllFieldsInPf.GenWithStackByArgs("PRIMARY KEY")
+				}
+				return ErrUniqueKeyNeedAllFieldsInPf.GenWithStackByArgs("UNIQUE INDEX")
```
Comment (crazycs520) [inline ddl/partition.go]:
```sql
create table part7 (
		col1 int not null,
		col2 int not null,
		col3 int not null,
		unique key(col1, col2),
		primary key (col3)
	)
	partition by range (col1 + col2) (
	partition p1 values less than (11),
	partition p2 values less than (15)
	)
```
MySQL
```
(1503, u"A PRIMARY KEY must include all columns in the table's partitioning function")
```

This PR
``` 
(1503, u"A UNIQUE INDEX must include all columns in the table's partitioning function")
```
and execute below sql got the same result too, but i can't understand the mysql behavier
```sql
create table part7 (
		col1 int not null,
		col2 int not null,
		col3 int not null unique,
		unique key(col1, col2)
	)
	partition by range (col1 + col2) (
	partition p1 values less than (11),
	partition p2 values less than (15)
	)
```

File: ddl/partition.go [hunk 3]
```diff
@@ -435,56 +435,36 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
...
-			// Extract every unique key and primary key.
-			if len(uniKeys) != 0 {
-				constraints = append(constraints, uniKeys)
+	// Every unique key on the table must use every column in the table's partitioning expression.
+	// See https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html
+	for _, constraint := range constraints {
+		switch constraint.Tp {
+		case ast.ConstraintPrimaryKey, ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex:
+			if !checkUniqueKeyIncludePartKey(partCols, constraint.Keys) {
+				if constraint.Tp == ast.ConstraintUniq {
```
Comment (zimulala) [inline ddl/partition.go]: Why not use `ConstraintPrimaryKey` here?

Comment (zimulala) [inline ddl/partition.go]: Why not use `ConstraintPrimaryKey`?

**Comment by tiancaiamao** [comment]
PTAL @zimulala @crazycs520
/run-all-tests

File: ddl/partition.go [hunk 4]
```diff
@@ -494,42 +474,41 @@ func checkPartitionKeysConstraint(sctx sessionctx.Context, partExpr string, idxC
...
+func checkUniqueKeyIncludePartKey(partCols []*expression.Column, idxCols []*ast.IndexColName) bool {
+	for _, partCol := range partCols {
+		if !findColumnInColumns(partCol, idxCols) {
 			return false
 		}
 	}
 	return true
 }
 
+func findColumnInColumns(c *expression.Column, cols []*ast.IndexColName) bool {
```
Comment (zimulala) [inline ddl/partition.go]: s/findColumnInColumns/findColumnInIndexCols ?
Could we put it to ddl/column.go ?

Comment (tiancaiamao) [reply ddl/partition.go]: Done

**Pull request merged**